src/SimpleParallelAnalyzer.cpp

```diff
@@ -3,160 +3,210 @@
+            }
+            mResults->AddMarker( sample, AnalyzerResults::Dot, mDataChannels[ i ] );
+        }
+
+
+        Frame frame;
+        frame.mData1 = result;
+        frame.mFlags = 0;
+
+        // The code in these if/else blocks could be replaced with to `AdvanceToNextEdge` calls, but if no more transitions are encountered,
+        // the current state will never be output as a frame. These blocks will detect that case in the available data, and output a frame
+        // immediately, and then another one once the next active edge sample is known.
+        if( !mClock->DoMoreTransitionsExistInCurrentData() )
+        {
+            // There are no transitions available in the current data. Estimate what the frame size might be as 10 if we haven't seen any
+            // frames yet, otherwise use 10% of the last frame size.
+            U64 estimated_frame_size = 10;
+            if( !is_first_frame )
+            {
+                U64 last_frame_sample_count = last_frame.mEndingSampleInclusive - last_frame.mStartingSampleInclusive + 1;
+                estimated_frame_size = last_frame_sample_count * 0.1;
+                if( estimated_frame_size == 0 )
+                {
+                    estimated_frame_size = 1;
+                }
+            }
+
+            // Make sure we haven't gone past the end of the _real_ frame. WouldAdvancingCauseTransition can block, in which case more
+            // transitions might show up.
+            if( mClock->WouldAdvancingCauseTransition( estimated_frame_size ) )
+            {
+                // Move to inactive edge
+                mClock->AdvanceToNextEdge();
+                if( !mClock->DoMoreTransitionsExistInCurrentData() )
+                {
+                    frame.mStartingSampleInclusive = sample;
+                    frame.mEndingSampleInclusive = mClock->GetSampleNumber() - 1;
+                    mResults->AddFrame( frame );
+                    mResults->CommitResults();
+
+                    sample = mClock->GetSampleNumber();
+                }
+            }
+            else
+            {
+                frame.mStartingSampleInclusive = sample;
+                frame.mEndingSampleInclusive = sample + estimated_frame_size - 1;
+                mResults->AddFrame( frame );
+                mResults->CommitResults();
+
+                sample += estimated_frame_size;
+
+                // Move to inactive edge
+                mClock->AdvanceToNextEdge();
+            }
+
+            // Move to active edge
+            mClock->AdvanceToNextEdge();
+        }
+        else
+        {
+            // Move to inactive edge
+            mClock->AdvanceToNextEdge();
+
+            if( !mClock->DoMoreTransitionsExistInCurrentData() )
+            {
+                frame.mStartingSampleInclusive = sample;
+                frame.mEndingSampleInclusive = mClock->GetSampleNumber() - 1;
+                mResults->AddFrame( frame );
```

Comment:
is it possible if the frame gets double added here, once here, amd then again at the bottom of the loop?
This would require DoMoreTransitionsExistInCurrentData() to return false here, but then have AdvanceToNextEdge not block forever, which would happen if there was a large gap in the data here.

Reply:
The sample gets moved forward below. I think this is right, but maybe there is a case I am missing.